src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Tigase XMPP Server - The instant messaging server
+ * Copyright (C) 2004 Tigase, Inc. (office@tigase.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, version 3 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. Look for COPYING file in the top folder.
+ * If not, see http://www.gnu.org/licenses/.
+ */
+
+package tigase.xmpp.impl;
+
+import tigase.auth.*;
+import tigase.db.AuthRepository;
+import tigase.db.TigaseDBException;
+import tigase.kernel.beans.Inject;
+import tigase.server.Command;
+import tigase.server.Packet;
+import tigase.server.Priority;
+import tigase.xml.Element;
+import tigase.xmpp.NotAuthorizedException;
+import tigase.xmpp.StanzaType;
+import tigase.xmpp.XMPPProcessorIfc;
+import tigase.xmpp.XMPPResourceConnection;
+import tigase.xmpp.jid.BareJID;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+abstract public class SaslAuthAbstract
+		extends AbstractAuthPreprocessor
+		implements XMPPProcessorIfc {
+
+	protected final static String SASL_SERVER_KEY = "SASL_SERVER_KEY";
+	protected final static String ALLOWED_SASL_MECHANISMS_KEY = "allowed-sasl-mechanisms";
+
+	private static final Logger log = Logger.getLogger(SaslAuthAbstract.class.getName());
+
+	public enum ElementType {
```

Comment:
I'm not sure we should "merge" two sets of element types for 2 different SASL protocols into one enum (set/collection). I think any change in the name of elements in SASL2 (which is still possible) can result in a lot of errors as it may require updating in some places but not others. I would prefer to keep the protocol part separate from the logic part, so we could "merge" logic into one code, but keep 2 separate protocol implementations.

Reply:
My reasoning was: it's still SASL so within same area. What's more - this is the list of possible Element types - I'm not sure that those should/would change, rather new would be added but addition wouldn't case issues with logic/how it's handled.

Reply:
There are 2 SASLs, different versions, so they can be different. I'm not keen on relaying on the fact "they are similar" currently.

Reply:
Well, there are 3 SASLs: old one, new one and proposed changes to new one...